### ISSUE TITLE
Workaround and data collection for multiple-text-select deletion

### ISSUE THREAD

**Issue by davidskalinder** [davidskalinder/mpeds-coder#92]
Done in 6098fb7; merged into testing branch and live in the testing deployment.  If testers think this is okay, needs merge to master and production deployment as well as a PR upstream.

**Comment by davidskalinder** [comment]
So, @olderwoman and @matthewsmorgan:

There are several options for how I think we could fix #89 in the long term, which I'll outline in that thread.  But for the short term I think the best thing to do is to force the workaround that we've asked the coders to use (i.e., to not delete multiply-entered selected text), in case they forget or didn't get the message.  (Pushing this workaround upstream will also give @alexhanna a quick-and-easy way to mitigate the problem.)

So I've added functionality for the server to refuse requests to delete multiple text-select records at once, and to raise an error message in the UI that says "Error deleting item. Leave duplicates in so we can collect data on this bug."  That should stabilize the situation until we get a real fix in place, and ensure we get the data to measure how bad the problem was.

I think it does need a test from some people who are used to using the UI though -- would you mind taking a look at the testing deployment and see how it goes when you add text-selects multiple times and then delete them?  Remember, if you want to see the real state of what's in the database, you can refresh the page (changing tabs in the coder panel won't do the trick).

Let me know if you need any more info about how to test.  I think it'd be good to get this workaround live (and sent upstream) ASAP, so if you think the functionality is good enough, let me know and I'll move it into production.

**Comment by davidskalinder** [comment]
@olderwoman / @matthewsmorgan, just wanted to send you a quick reminder about this one since despite the useful and interesting discussion over at issue #89, this one is actually the higher priority since it'll hopefully stop the bleeding entirely (even beyond the considerable slowing that will likely result from our email to the coders).

However I don't think I should make the change live until you two make sure it's okay, so please let me know if you need any more info in order to test it out.

**Comment by olderwoman** [comment]
Sorry I failed to spot that you wanted quick turn-around on error testing. You can't do it? I'll log into the testing deployment now and give it 10 minutes.

**Comment by olderwoman** [comment]
One test in testing version. 1#. I marked the same text twice.  Then deleted one of them, it asked me was I sure, I said yes, then it refreshed the page with both in place. I did not see the error message until after I  hit “save” and returned to the article-level page,, not the event-level page. So the error message appears in the wrong place and I can imagine a coder getting very frustrated about why the deletion didn’t work and continuing to try it, since the relevant error message does not appear at that point, instead you get the “are  you sure you want to delete this?” message.

It seems like this behavior would have been detected if there had been any test at all of the user experience? The “error deleting item” message would need to appear in place of the “are you sure you want to delete?” message.

**Comment by olderwoman** [comment]
More info, coming back into the event page after a save, attempt to delete the duplicate simply fails, but there is no error message. Clicking does nothing.
Trying again:   OK I did screen shots this time, I now see that the error message does appear when I delete, but it is nowhere near where I was working when I did the deletion. See screen shots.

**Comment by johnklemke** [comment]
As I'm understanding this, I think it won't be too hard to handle this in the code I intend to write to substitute for an append query. David, we talked about this a bit at Monday's meeting, but let me recap (and open up for questions from others). I see the process of bringing the latest data from MAI into Access in a form ready for pass 2 as a two step process. First, we'll use a saved import specification to grab data from a .csv file and drop it into an Access table, doing a minimum of conversion, mostly taking the fields of interest into short or long text fields. Since the .csv file will always be a complete redump of all the MAI data, each time it gets imported into an Access table it'll go into a fresh version of the table, the old version deleted. One thing I like about this .csv to Access table process is that if there are errors, Access creates a table that logs them referencing row numbers in the .csv. But to make the data usable for the Access forms that are the guts of the Pass 2 process, the short and long text fields in the staging table need to get converted into the data types needed in the coder lines table that's part of the guts of the Pass 2 process. That's the second step. But I don't see a way to configure Access to run an append query that logs conversion errors like the import process does. It just reports that something like 59 fields could not be converted and the intended values were set to null. No freakin' bread crumbs, no handy way for a non-programmer to resolve the issues. So my current plan -- which I may well start on tomorrow -- is to run the equivalent of an append query in procedural code written in VBA. That way at each step of the process, I'll know what record from the staging table is in hand, and what record in the growing coder lines table is the target. When a special case (like a duplicate value) is detected, or when a conversion error occurs, I'll be able to log it into an ancillary table for logging any errors and weirdnesses in the append process. Bottom line -- yeah, I think we can do what's needed in the set up for pass 2, and I already had reasons to build the core of the code anyhow.

**Comment by matthewsmorgan** [comment]
Seeing this now. It sounds like this is mostly a question for John, who has
to implement this in Pass 2. I'll work with whatever I'm given in the
Access forms (frankly I didn't use the text select info a lot last time
because a lot of the old data didn't have that anyway).

**Comment by olderwoman** [comment]
Duplicated text has an extraordinarily small likelihood of being a problem in pass 2, in fact I think that likelihood is indistinguishable from zero. Pass 2 coder experiences mild annoyance at seeing the same text twice in a field. They can probably live through the experience with little loss of time or efficiency unless they forget it is a known bug and worry about needing to report the problem.

**Comment by davidskalinder** [comment]
Okay, sounds like this won't be much of a problem then.  To clarify, I don't think this will really need any new pass 2 functionality (since I think pass 2 reviewers should already be able to see when a variable has multiple text-select entries?).  But this change will make it impossible for pass 1 coders to delete duplicate entries even if they are mistakes, so all duplicate entries will have to be treated as possible mistakes in pass 2.

All that said, it sounds like I'm more worried about this than anyone else lol.  So I'll go ahead and make this workaround live.  Then I think we should give it a couple weeks in order to see how often duplicates occur before implementing a fix for #89, which will override the workaround.

**Comment by davidskalinder** [comment]
All right, this is now live in the production deployment.  I haven't re-tested it there because I didn't want to create bad data, but it should be identical to what's in the testing deployment (and I can confirm that it doesn't break the main coding interface).

I'm going to submit a Pull Request so that @alexhanna and the gang can use this as well, but otherwise I think this is done.  Thanks testing team!

**Comment by davidskalinder** [comment]
I'm reopening this since I realized I forgot about the data collection part.  Here's the query that I think gives us the data we want:

```
/* Number of multiple entries of text-capture variables */
SELECT coder_id
  , event_id
  , variable
  , value
  , COUNT(id) AS n
  FROM coder_event_creator
  WHERE RIGHT(variable, 5) = "-text"
  AND timestamp > "2020-06-11 12:00"
  GROUP BY coder_id
  , event_id
  , variable
  , value
  , text
  HAVING n > 1
  ;
```

... and the results from our production database since the change went live 18 days ago:

```
+----------+----------+----------------+-----------+---+
| coder_id | event_id | variable       | value     | n |
+----------+----------+----------------+-----------+---+
|       14 |     6085 | named-org-text | 7-207-255 | 2 |
|       14 |     6133 | actor-text     | 0-118-143 | 2 |
|       14 |     6133 | location-text  | 0-61-86   | 2 |
|       17 |     6559 | time-text      | 1-160-170 | 2 |
+----------+----------+----------------+-----------+---+
4 rows in set (0.09 sec)
```

For context:
```
mysql> SELECT COUNT(*) FROM event_creator_queue WHERE coded_dt > "2020-06-11 12:00";
+----------+
| COUNT(*) |
+----------+
|      376 |
+----------+
1 row in set (0.00 sec)
```

So yeah, this has been awfully rare since we made this change: just 4 times out of 376 articles coded!  And actually, 3 of those were in the first few hours after I made the change, so it's possible that coder 14 was just testing the new functionality.

However, we did tell our coders to watch out for this, so maybe they were being more careful?  Peeking into the campus protests DB, they've had it happen a little more frequently since roughly when @alexhanna made the fix live: 10 times in about 395 articles, which is about 2.5%.

@alexhanna, did you tell your coders about this problem, or did you just accept the PR silently?  If you did it silently, then we can probably draw a confidence interval around that 2.5% number and expect that about that many articles from the past are missing a piece of text-select data.  If you told them, then I guess we should adjust that expectation somehow...

**Comment by davidskalinder** [comment]
Just for my reference, in case I need it later and since it's a pain to write out all those fields, here's the query to pull each row with a duplicate entry:

```
/* Rows with text-capture multi-entry variables */
SELECT a.id
  , a.coder_id
  , a.event_id
  , a.variable
  , a.value
  , a.timestamp
  FROM coder_event_creator AS a
  INNER JOIN
    (SELECT coder_id
       , event_id
       , variable
       , value
       , text
       , COUNT(id) AS n
       FROM coder_event_creator
       WHERE RIGHT(variable, 5) = "-text"
       GROUP BY coder_id
       , event_id
       , variable
       , value
       , text
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.event_id = b.event_id
  AND a.variable = b.variable
  AND a.value = b.value
  AND a.text = b.text
  ;
```

**Comment by alexhanna** [comment]
I did not tell them about this issue.

**Comment by davidskalinder** [comment]
> I did not tell them about this issue.

Hooray for methodological rigor!

So using your clean sample, there are 10 occurrences out of 395 articles.  Since these are counts, I think this is a Poisson distribution?  And apparently calculating CIs for Poisson isn't [all that well defined](https://stats.stackexchange.com/questions/15371/how-to-calculate-a-confidence-level-for-a-poisson-distribution) unless n is big enough to pretend it's normal, which I think 395 should be... so that lets us use `lambda +- 1.96 * sqrt(lambda / n)`, where `lambda` is the observed mean.  I get from that a CI from about .0096 to .0410; so unless this sample is part of the unlucky 5%, lambda is between about .01 and .04.  Methods and Stats prelim prep FTW!

So, tl;dr: we've probably been losing

**Comment by davidskalinder** [comment]
Looking back at old campus protest data, it looks like these double-entry errors were almost always corrected by coders before the fix went in: there's only about one occurrence a month.  That is bad, of course, since what the coders thought was a correction was actually a deletion.

So we should probably assume that that 1-4% number is a good estimate of the number of articles coded prior to early June 2020 that are erroneously missing a text-select entry.  (Well, more strictly speaking, there are probably about .01-.04 text-selects missing per article.)

I'm not sure what the actions to be taken from here are, so I'll leave the issue open in case others (@olderwoman / @limchaeyoon / @matthewsmorgan ?) want to weigh in.  But now I think we can worry about fixing #89 properly with a pretty good idea of how it's gone wrong in the past.

**Comment by davidskalinder** [comment]
A clarification from @alexhanna on this one: she did in fact inform the campus protest coders about this issue, so they created about .025 duplicate text-selects per article even knowing that it was a bad thing to do.  (Not sure why that's so much higher than our rate under similar conditions, but hey.)

I can't remember whether I checked to see if the campus protesters' duplicates looked like intentional testing, so I'll do that.  If they do look natural, then maybe that 1-4% should be viewed as a lower bound of what would happen without foreknowledge?

**Comment by davidskalinder** [comment]
> I can't remember whether I checked to see if the campus protesters' duplicates looked like intentional testing, so I'll do that.

So these occurrences do indeed look like they are natural: the ten occurrences happened across four users, with hours between each occurrence.

Given that the campus protest coders were working under the same conditions as ours, I guess it makes the most sense to pool the samples for a rate of 14 occurrences over 771 articles; using the formula above that gets us to a CI of .013-.023.  So perhaps that's slightly more optimistic, although we still don't know the rate for coders who aren't watching out for this problem.

**Comment by davidskalinder** [comment]
Outcome of the team discussion on this: "meh".  Other coder errors almost certainly swamp this bug, and pass 2 coders have to reconcile them all anyway.  So the plan for this is to fix it in #89 and move on with our lives.

Closing now.
